File: tests/test_python_multiple_inheritance.py
# Adapted from:
# https://github.com/google/clif/blob/5718e4d0807fd3b6a8187dde140069120b81ecef/clif/testing/python/python_multiple_inheritance_test.py

import pytest

from pybind11_tests import python_multiple_inheritance as m


class PC(m.CppBase):
    pass


class PPCC(PC, m.CppDrvd):
    pass


class PPPCCC(PPCC, m.CppDrvd2):
    pass


class PC1(m.CppDrvd):
    pass


class PC2(m.CppDrvd2):
    pass


class PCD(PC1, PC2):
    pass


class PCDI(PC1, PC2):
    def __init__(self):
        PC1.__init__(self, 11)
        PC2.__init__(self, 12)


def test_PC():
    d = PC(11)
    assert d.get_base_value() == 11
    d.reset_base_value(13)
    assert d.get_base_value() == 13


def test_PPCC():
    d = PPCC(11)
    assert d.get_drvd_value() == 33
    d.reset_drvd_value(55)
    assert d.get_drvd_value() == 55

    assert d.get_base_value() == 11
    assert d.get_base_value_from_drvd() == 11
    d.reset_base_value(20)
    assert d.get_base_value() == 20
    assert d.get_base_value_from_drvd() == 20
    d.reset_base_value_from_drvd(30)
    assert d.get_base_value() == 30
    assert d.get_base_value_from_drvd() == 30


def NOtest_PPPCCC():
    # terminate called after throwing an instance of 'pybind11::error_already_set'
    # what():  TypeError: bases include diverging derived types:
    #     base=pybind11_tests.python_multiple_inheritance.CppBase,
    #     derived1=pybind11_tests.python_multiple_inheritance.CppDrvd,
    #     derived2=pybind11_tests.python_multiple_inheritance.CppDrvd2
    PPPCCC(11)


def test_PCD():
    # This escapes all_type_info_check_for_divergence() because CppBase does not appear in bases.
    with pytest.raises(
        TypeError,
        match=r"CppDrvd2\.__init__\(\) must be called when overriding __init__$",
    ):
        PCD(11)


def test_PCDI():
    obj = PCDI()
    with pytest.raises(TypeError, match="^bases include diverging derived types: "):
        m.pass_CppBase(obj)

File: tests/test_python_multiple_inheritance.cpp
#include "pybind11_tests.h"

namespace test_python_multiple_inheritance {

// Copied from:
// https://github.com/google/clif/blob/5718e4d0807fd3b6a8187dde140069120b81ecef/clif/testing/python_multiple_inheritance.h

struct CppBase {
    explicit CppBase(int value) : base_value(value) {}
    int get_base_value() const { return base_value; }
    void reset_base_value(int new_value) { base_value = new_value; }

private:
    int base_value;
};

struct CppDrvd : CppBase {
    explicit CppDrvd(int value) : CppBase(value), drvd_value(value * 3) {}
    int get_drvd_value() const { return drvd_value; }
    void reset_drvd_value(int new_value) { drvd_value = new_value; }

    int get_base_value_from_drvd() const { return get_base_value(); }
    void reset_base_value_from_drvd(int new_value) { reset_base_value(new_value); }

private:
    int drvd_value;
};

struct CppDrv2 : CppDrvd {
    explicit CppDrv2(int value) : CppDrvd(value), drv2_value(value * 5) {}
    int get_drv2_value() const { return drv2_value; }

private:
    int drv2_value;
};

} // namespace test_python_multiple_inheritance

TEST_SUBMODULE(python_multiple_inheritance, m) {
    using namespace test_python_multiple_inheritance;

    py::class_<CppBase>(m, "CppBase")
        .def(py::init<int>())
        .def("get_base_value", &CppBase::get_base_value)
        .def("reset_base_value", &CppBase::reset_base_value);

    py::class_<CppDrvd, CppBase>(m, "CppDrvd")
        .def(py::init<int>())
        .def("get_drvd_value", &CppDrvd::get_drvd_value)
        .def("reset_drvd_value", &CppDrvd::reset_drvd_value)
        .def("get_base_value_from_drvd", &CppDrvd::get_base_value_from_drvd)
        .def("reset_base_value_from_drvd", &CppDrvd::reset_base_value_from_drvd);

#if 1 // This works.
    py::class_<CppDrv2, CppDrvd>(m, "CppDrv2")
#else // Apparent undefined behavior.
    py::class_<CppDrv2, CppDrvd, CppBase>(m, "CppDrv2")
#endif
        .def(py::init<int>())
        .def("get_drv2_value", &CppDrv2::get_drv2_value);
}
